logstash-rs/src/buffer.rs

```diff
@@ -0,0 +1,174 @@
+}
+
+impl BufferedSender {
+    pub fn new<S: Sender>(
+        sender: S,
+        buffer_size: Option<usize>,
+        buffer_lifetime: Option<Duration>,
+        ignore_buffer: Level,
+    ) -> Self {
+        let sender =
+            BufferedSenderThread::new(sender, buffer_size, buffer_lifetime, ignore_buffer).run();
+        Self { sender }
+    }
+}
+
+impl Sender for BufferedSender {
+    fn send(&self, event: LogStashRecord) -> Result<()> {
+        self.sender.send(Command::Send(event))?;
+        Ok(())
+    }
+
+    fn send_batch(&self, events: Vec<LogStashRecord>) -> Result<()> {
+        self.sender.send(Command::SendBatch(events))?;
+        Ok(())
+    }
+
+    fn flush(&self) -> Result<()> {
+        self.sender.send(Command::Flush)?;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct BufferedSenderThread<S: Sender> {
+    sender: S,
+    buffer: Vec<LogStashRecord>,
+    buffer_size: Option<usize>,
+    buffer_lifetime: Option<Duration>,
+    deadline: Option<Instant>,
+    ignore_buffer: Level,
+}
+
+impl<S: Sender> BufferedSenderThread<S> {
+    fn new(
+        sender: S,
+        buffer_size: Option<usize>,
+        buffer_lifetime: Option<Duration>,
+        ignore_buffer: Level,
+    ) -> Self {
+        let buffer_size = match buffer_size {
+            Some(s) if s < 2 => None,
+            x => x,
+        };
+        Self {
+            sender,
+            buffer: Vec::with_capacity(buffer_size.unwrap_or(0)),
+            buffer_size,
+            buffer_lifetime,
+            deadline: None,
+            ignore_buffer,
+        }
+    }
+
+    fn run(self) -> mpsc::SyncSender<Command> {
+        let (sender, receiver) = mpsc::sync_channel(1);
+        self.run_thread(receiver);
+        sender
+    }
+
+    fn next_deadline(&self) -> Option<Instant> {
+        if self.buffer.is_empty() && self.buffer_size.is_some() {
+            return self.buffer_lifetime.map(|lt| Instant::now() + lt);
+        }
+        None
+    }
+
+    fn run_thread(mut self, receiver: mpsc::Receiver<Command>) {
+        std::thread::spawn::<_, Result<()>>(move || {
+            {
+                loop {
+                    let cmd = match self.deadline {
+                        Some(deadline) => receiver
+                            .recv_timeout(deadline.saturating_duration_since(Instant::now())),
+                        None => receiver
+                            .recv()
+                            .map_err(|_| mpsc::RecvTimeoutError::Disconnected),
+                    };
+
+                    if let Ok(Command::SendBatch(_) | Command::Send(_)) = &cmd {
+                        self.deadline = self.next_deadline();
+                    }
+                    match cmd {
+                        Ok(Command::Flush) | Err(mpsc::RecvTimeoutError::Timeout) => {
+                            self.flush()?
+                        }
+                        Ok(Command::Send(event)) => self.send(event)?,
```

Comment:
You are propagating the error from `send_bytes` up to this point. Do I understand correctly that the thread will stop if the TcpStream at the lowest level observe a disconnect? If yes, then it needs to be fixed